modal/runner/engines/vllm.py

```diff
@@ -70,84 +84,82 @@ def __init__(self, params: VllmParams):
-                    finish_reason=finish_reason,
-                )
-            else:
-                yield create_response_text(
-                    output,
-                    prompt_tokens,
-                    completion_tokens,
-                    done=True,
-                    finish_reason=finish_reason,
-                )
-
-            elapsed = time.perf_counter() - t_start_inference
+            async for current in results_generator:
+                output = current.outputs[0].text
+                finish_reason = current.outputs[0].finish_reason
+                resp.usage.prompt_tokens = len(current.prompt_token_ids)
+                resp.usage.completion_tokens = len(current.outputs[0].token_ids)
+                resp.usage.duration = time.time() - t_start_usage_duration
+
+                # Non-streaming requests continue generating w/o yielding intermediate results
+                if not payload.stream:
+                    yield " "  # HACK: Keep the connection alive while generating
+                    continue
+
+                # Skipping invalid UTF8 tokens:
+                if output and output[-1] == "\ufffd":
+                    continue
+
+                # Streaming requests send SSE messages with each new generated part
+                token = output[index:]
+                index = len(output)
+                resp.text = token
+                resp.finish_reason = finish_reason
+                yield sse(resp.json(ensure_ascii=False))
+
+            resp.text = "" if payload.stream else output
+            resp.done = True
+            resp.finish_reason = finish_reason
+            data = resp.json(ensure_ascii=False)
+            yield sse(data) if payload.stream else data
+
             logger.info(
                 "Completed generation",
                 extra={
                     "model": self.engine_args.model,
-                    "tokens": completion_tokens,
-                    "tps": completion_tokens / elapsed,
-                    "duration": elapsed,
+                    "tokens": resp.usage.completion_tokens,
+                    "tps": resp.usage.completion_tokens / t_start_inference,
+                    "duration": resp.usage.duration,
                 },
             )
         except Exception as err:
             e = create_error_text(err)
             logger.exception(
                 "Failed generation", extra={"model": self.engine_args.model}
             )
-            if payload.stream:
-                yield create_sse_data(e)
-            else:
-                yield e
+            # TODO: Existing code prior to refactor was yielding error text
+            # in a ResponseBody model, with empty usage data. Non-streaming
+            # requests yielded only error text as a string. This behavior
+            # is preserved here, but should maybe be revisited.
+            resp.text = e
+            resp.usage.prompt_tokens = 0
+            resp.usage.completion_tokens = 0
```

Comment:
Feel like we don't have to reset it here. The exception could have been thrown mid generation :-?

Reply:
Ah, actually we should just yield sse(e) instead of a ResponBody with text = e, since this is create_error_text:

```python
def create_error_text(err: Exception) -> str:
    return ErrorResponse(
        error=ErrorPayload(message=f"{err}", type=f"{type(err).__name__}")
    ).json(ensure_ascii=False)
```

It's basically creating an ErrorResponse shape so that OR web can handle. So we shouldn't need the lines above at all I think.

Reply:
The error payload for SSE vs non-SSE is the same, just that one needed to be wrapped inside the sse prefix!

Reply:
sick! removed here e9d7e2d